library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/RtspMediaSource.java

```diff
@@ -0,0 +1,225 @@
+            isCreateCalled = true;
+            return new RtspMediaSource(
+                    uri,
+                    factory,
+                    minLoadableRetryCount,
+                    treatLoadErrorsAsEndOfStream);
+        }
+
+        /**
+         * Returns a new {@link RtspMediaSource} using the current parameters. Media source events
+         * will not be delivered.
+         *
+         * @param uri The {@link Uri}.
+         * @param eventHandler A handler for events.
+         * @param eventListener A listener of events.
+         * @return The new {@link RtspMediaSource}.
+         */
+        public RtspMediaSource createMediaSource(Uri uri,
+                                                 @Nullable Handler eventHandler,
+                                                 @Nullable MediaSourceEventListener eventListener) {
+            RtspMediaSource mediaSource = createMediaSource(uri);
+            if (eventHandler != null && eventListener != null) {
+                mediaSource.addEventListener(eventHandler, eventListener);
+            }
+            return mediaSource;
+        }
+
+    }
+
+
+    /**
+     * The default minimum number of times to retry loading data prior to failing.
+     */
+    private static final int DEFAULT_MIN_LOADABLE_RETRY_COUNT = 3;
+
+    private final Uri uri;
+    private final Client.Factory<? extends Client> factory;
+    private final int minLoadableRetryCount;
+    private final boolean treatLoadErrorsAsEndOfStream;
+    private EventDispatcher eventDispatcher;
+
+    private Client client;
+
+    private RtspMediaSource(Uri uri, Client.Factory<? extends Client> factory,
+                           int minLoadableRetryCount,
+                           boolean treatLoadErrorsAsEndOfStream) {
+        this.uri = uri;
+        this.factory = factory;
+        this.minLoadableRetryCount = minLoadableRetryCount;
+        this.treatLoadErrorsAsEndOfStream = treatLoadErrorsAsEndOfStream;
+    }
+
+    // MediaTrackSource implementation
+
+    @Override
+    public void maybeThrowSourceInfoRefreshError() throws IOException {
+        if (client == null) {
+            throw new IOException();
+        }
+    }
+
+    @Override
+    public MediaPeriod createPeriod(MediaPeriodId id, Allocator allocator) {
+        Assertions.checkArgument(id.periodIndex == 0);
+        eventDispatcher = createEventDispatcher(id);
+        return new RtspMediaPeriod(client.session(), minLoadableRetryCount, allocator);
+    }
+
+    @Override
+    public void releasePeriod(MediaPeriod mediaPeriod) {
+        ((RtspMediaPeriod) mediaPeriod).release();
+    }
+
+
+    @Override
+    public void prepareSourceInternal(ExoPlayer player, boolean isTopLevelSource) {
+        client = new Client.Builder(factory).uri(uri).listener(this).build();
+
+        try {
+
+            player.getVideoComponent().addVideoListener(client.session());
+            client.open();
+
+        } catch (NullPointerException e) {
+
```

Comment:
We shouldn't silently consume exceptions without callbacks informing of the failed state, etc. At the very least we should be logging the exception so that library consumers know what happened to help track down issues.